# 1 Задание

### Импорт данных

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('task/2016-FCC-New-Coders-Survey-Data.csv', low_memory=False)
df.head()

Age  AttendedBootcamp  BootcampFinish  BootcampFullJobAfter  \
0  28.0               0.0             NaN                   NaN   
1  22.0               0.0             NaN                   NaN   
2  19.0               0.0             NaN                   NaN   
3  26.0               0.0             NaN                   NaN   
4  20.0               0.0             NaN                   NaN   

   BootcampLoanYesNo  BootcampMonthsAgo BootcampName  BootcampPostSalary  \
0                NaN                NaN          NaN                 NaN   
1                NaN                NaN          NaN                 NaN   
2                NaN                NaN          NaN                 NaN   
3                NaN                NaN          NaN                 NaN   
4                NaN                NaN          NaN                 NaN   

   BootcampRecommend  ChildrenNumber  ... ResourceSoloLearn  \
0                NaN             NaN  ...               NaN   
1                NaN             NaN  ...               NaN   
2                NaN             NaN  ...               NaN   
3                NaN             NaN  ...               NaN   
4                NaN             NaN  ...               NaN   

   ResourceStackOverflow  ResourceTreehouse  ResourceUdacity  ResourceUdemy  \
0                    NaN                NaN              NaN            NaN   
1                    NaN                NaN              NaN            1.0   
2                    NaN                NaN              NaN            NaN   
3                    NaN                NaN              NaN            NaN   
4                    NaN                NaN              NaN            NaN   

   ResourceW3Schools  ResourceYouTube  \
0                NaN              NaN   
1                NaN              NaN   
2                NaN              NaN   
3                NaN              NaN   
4                NaN              NaN   

                              SchoolDegree              SchoolMajor  \
0           some college credit, no degree                      NaN   
1           some college credit, no degree                      NaN   
2  high school diploma or equivalent (GED)                      NaN   
3                        bachelor's degree  Cinematography And Film   
4           some college credit, no degree                      NaN   

   StudentDebtOwe  
0         20000.0  
1             NaN  
2             NaN  
3          7000.0  
4             NaN  

[5 rows x 113 columns]

# 2 Задание

### Выбор нужных стобцов

In [3]:
df1 = df.loc[:, ["CityPopulation", "EmploymentStatus", "Gender", "HasDebt", "JobPref", "JobWherePref", 
           "MaritalStatus", "Income", "SchoolDegree"]]
df1

CityPopulation          EmploymentStatus  Gender  \
0      between 100,000 and 1 million        Employed for wages    male   
1      between 100,000 and 1 million        Employed for wages    male   
2                more than 1 million        Employed for wages    male   
3                more than 1 million        Employed for wages  female   
4      between 100,000 and 1 million        Employed for wages  female   
...                              ...                       ...     ...   
15615            more than 1 million  Self-employed freelancer    male   
15616            more than 1 million        Employed for wages    male   
15617            more than 1 million  Self-employed freelancer    male   
15618            more than 1 million  Self-employed freelancer    male   
15619            more than 1 million        Employed for wages  female   

       HasDebt                          JobPref  \
0          1.0                       freelance    
1          0.0               work for a startup   
2          0.0          start your own business   
3          1.0               work for a startup   
4          1.0  work for a medium-sized company   
...        ...                              ...   
15615      1.0          start your own business   
15616      0.0               work for a startup   
15617      0.0                       freelance    
15618      0.0                              NaN   
15619      1.0                       freelance    

                             JobWherePref                    MaritalStatus  \
0                                     NaN  married or domestic partnership   
1      in an office with other developers                              NaN   
2                                     NaN                              NaN   
3                               from home                              NaN   
4      in an office with other developers                              NaN   
...                                   ...                              ...   
15615                                 NaN            single, never married   
15616  in an office with other developers            single, never married   
15617                                 NaN                              NaN   
15618                                 NaN                              NaN   
15619                                 NaN            single, never married   

        Income                             SchoolDegree  
0      32000.0           some college credit, no degree  
1      15000.0           some college credit, no degree  
2      48000.0  high school diploma or equivalent (GED)  
3      43000.0                        bachelor's degree  
4       6000.0           some college credit, no degree  
...        ...                                      ...  
15615  10000.0                        bachelor's degree  
15616  10000.0                        bachelor's degree  
15617  48000.0                        bachelor's degree  
15618  10000.0       master's degree (non-professional)  
15619  18000.0                        bachelor's degree  

[15620 rows x 9 columns]

# 3 Задание

### Очистка данных 

In [4]:
df1.dropna(inplace=True)
df1

CityPopulation          EmploymentStatus  Gender  \
20               more than 1 million        Employed for wages  female   
36     between 100,000 and 1 million  Self-employed freelancer    male   
59               more than 1 million        Employed for wages    male   
71               more than 1 million        Employed for wages    male   
72                 less than 100,000        Employed for wages    male   
...                              ...                       ...     ...   
15564              less than 100,000        Employed for wages    male   
15566            more than 1 million        Employed for wages    male   
15584            more than 1 million        Employed for wages    male   
15598              less than 100,000        Employed for wages    male   
15616            more than 1 million        Employed for wages    male   

       HasDebt                               JobPref  \
20         1.0       work for a medium-sized company   
36         1.0       work for a medium-sized company   
59         1.0       work for a medium-sized company   
71         1.0  work for a multinational corporation   
72         1.0       work for a medium-sized company   
...        ...                                   ...   
15564      1.0       work for a medium-sized company   
15566      0.0                    work for a startup   
15584      1.0  work for a multinational corporation   
15598      1.0       work for a medium-sized company   
15616      0.0                    work for a startup   

                             JobWherePref                    MaritalStatus  \
20     in an office with other developers  married or domestic partnership   
36     in an office with other developers  married or domestic partnership   
59     in an office with other developers  married or domestic partnership   
71                              from home  married or domestic partnership   
72                              from home  married or domestic partnership   
...                                   ...                              ...   
15564  in an office with other developers  married or domestic partnership   
15566                       no preference  married or domestic partnership   
15584  in an office with other developers  married or domestic partnership   
15598                           from home  married or domestic partnership   
15616  in an office with other developers            single, never married   

         Income                             SchoolDegree  
20     200000.0       master's degree (non-professional)  
36      60000.0  high school diploma or equivalent (GED)  
59      35000.0           some college credit, no degree  
71      56000.0           some college credit, no degree  
72      35000.0                        bachelor's degree  
...         ...                                      ...  
15564  200000.0  professional degree (MBA, MD, JD, etc.)  
15566  200000.0                        bachelor's degree  
15584  200000.0                        bachelor's degree  
15598  200000.0                       associate's degree  
15616   10000.0                        bachelor's degree  

[935 rows x 9 columns]

In [5]:
df1 = df1[(df1['Gender'] == 'male') | (df1['Gender'] == 'female')]
df1

CityPopulation          EmploymentStatus  Gender  \
20               more than 1 million        Employed for wages  female   
36     between 100,000 and 1 million  Self-employed freelancer    male   
59               more than 1 million        Employed for wages    male   
71               more than 1 million        Employed for wages    male   
72                 less than 100,000        Employed for wages    male   
...                              ...                       ...     ...   
15564              less than 100,000        Employed for wages    male   
15566            more than 1 million        Employed for wages    male   
15584            more than 1 million        Employed for wages    male   
15598              less than 100,000        Employed for wages    male   
15616            more than 1 million        Employed for wages    male   

       HasDebt                               JobPref  \
20         1.0       work for a medium-sized company   
36         1.0       work for a medium-sized company   
59         1.0       work for a medium-sized company   
71         1.0  work for a multinational corporation   
72         1.0       work for a medium-sized company   
...        ...                                   ...   
15564      1.0       work for a medium-sized company   
15566      0.0                    work for a startup   
15584      1.0  work for a multinational corporation   
15598      1.0       work for a medium-sized company   
15616      0.0                    work for a startup   

                             JobWherePref                    MaritalStatus  \
20     in an office with other developers  married or domestic partnership   
36     in an office with other developers  married or domestic partnership   
59     in an office with other developers  married or domestic partnership   
71                              from home  married or domestic partnership   
72                              from home  married or domestic partnership   
...                                   ...                              ...   
15564  in an office with other developers  married or domestic partnership   
15566                       no preference  married or domestic partnership   
15584  in an office with other developers  married or domestic partnership   
15598                           from home  married or domestic partnership   
15616  in an office with other developers            single, never married   

         Income                             SchoolDegree  
20     200000.0       master's degree (non-professional)  
36      60000.0  high school diploma or equivalent (GED)  
59      35000.0           some college credit, no degree  
71      56000.0           some college credit, no degree  
72      35000.0                        bachelor's degree  
...         ...                                      ...  
15564  200000.0  professional degree (MBA, MD, JD, etc.)  
15566  200000.0                        bachelor's degree  
15584  200000.0                        bachelor's degree  
15598  200000.0                       associate's degree  
15616   10000.0                        bachelor's degree  

[931 rows x 9 columns]

# 4 Задание

### пункт a (Проверка нормальности распределения дохода)

In [7]:
import scipy.stats as sts
import numpy as np

In [8]:
df2 = df1.copy()

In [9]:
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.Income)
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.8556976914405823, p_value 2.449405406991464e-28


#### Гипотеза H0 о нормальном распределении отклоняется p < 0.05

In [10]:
df2['log_income'] = np.log(df2.Income + 1)
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.log_income)
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.9735251665115356, p_value 5.584226223792621e-12


#### Гипотеза H0 о нормальном распределении отклоняется p < 0.05

In [11]:
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.log_income[:100])
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.9772992730140686, p_value 0.0817166194319725


#### Гипотеза H0 о нормальном распределении принимается p > 0.05

### пункт b, c

In [13]:
import statsmodels.api as sm
from statsmodels.formula.api import ols 

In [44]:
df3 = df2[:100] 

## log_income - SchoolDegree

#### Результаты анализа ANOVA

In [115]:
model =  ols("log_income ~ C(SchoolDegree)", df3).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.062
Model:                            OLS   Adj. R-squared:                 -0.031
Method:                 Least Squares   F-statistic:                    0.6649
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.738
Time:                        00:54:09   Log-Likelihood:                -103.19
No. Observations:                 100   AIC:                             226.4
Df Residuals:                      90   BIC:                             252.4
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      10.5784      0.358     29.558      0.000       9.867      11.289
C(SchoolDegree)[T.associate's degree]                          -0.5535      0.449     -1.234      0.220      -1.445       0.338
C(SchoolDegree)[T.bachelor's degree]                            0.0928      0.378      0.246      0.807      -0.658       0.843
C(SchoolDegree)[T.high school diploma or equivalent (GED)]     -0.1375      0.438     -0.314      0.754      -1.008       0.733
C(SchoolDegree)[T.master's degree (non-professional)]          -0.0493      0.396     -0.125      0.901      -0.835       0.737
C(SchoolDegree)[T.no high school (secondary school)]            0.5037      0.800      0.629      0.531      -1.086       2.094
C(SchoolDegree)[T.professional degree (MBA, MD, JD, etc.)]      0.0968      0.438      0.221      0.826      -0.774       0.968
C(SchoolDegree)[T.some college credit, no degree]               0.0135      0.413      0.033      0.974      -0.808       0.834
C(SchoolDegree)[T.some high school]                            -0.1870      0.547     -0.342      0.733      -1.273       0.899
C(SchoolDegree)[T.trade, technical, or vocational training]     0.0552      0.506      0.109      0.913      -0.950       1.061
==============================================================================
Omnibus:                        1.321   Durbin-Watson:                   1.865
Prob(Omnibus):                  0.517   Jarque-Bera (JB):                1.297
Skew:                          -0.268   Prob(JB):                        0.523
Kurtosis:                       2.846   Cond. No.                         18.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.87 -> 2 => Автокорреляция отсутствует 

## log_income - CityPopulation

#### Результаты анализа ANOVA

In [116]:
model1 =  ols("log_income ~ C(CityPopulation)", df3).fit()
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     1.072
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.347
Time:                        00:54:15   Log-Likelihood:                -105.31
No. Observations:                 100   AIC:                             216.6
Df Residuals:                      97   BIC:                             224.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                   10.6430      0.126     84.138      0.000      10.392      10.894
C(CityPopulation)[T.less than 100,000]       0.0130      0.184      0.071      0.944      -0.351       0.377
C(CityPopulation)[T.more than 1 million]    -0.2032      0.168     -1.212      0.228      -0.536       0.129
==============================================================================
Omnibus:                        1.661   Durbin-Watson:                   1.822
Prob(Omnibus):                  0.436   Jarque-Bera (JB):                1.290
Skew:                          -0.273   Prob(JB):                        0.525
Kurtosis:                       3.103   Cond. No.                         3.89
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.82 -> 2 => Автокорреляция отсутствует 

## log_income - EmploymentStatus

#### Результаты анализа ANOVA

In [117]:
model2 =  ols("log_income ~ C(EmploymentStatus)", df3).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.044
Model:                            OLS   Adj. R-squared:                  0.024
Method:                 Least Squares   F-statistic:                     2.209
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.115
Time:                        00:54:23   Log-Likelihood:                -104.18
No. Observations:                 100   AIC:                             214.4
Df Residuals:                      97   BIC:                             222.2
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=======================================================================================================================
                                                          coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------------------------
Intercept                                              10.5568      0.074    142.216      0.000      10.409      10.704
C(EmploymentStatus)[T.Self-employed business owner]     0.5506      0.320      1.720      0.089      -0.085       1.186
C(EmploymentStatus)[T.Self-employed freelancer]        -0.3001      0.273     -1.097      0.275      -0.843       0.243
==============================================================================
Omnibus:                        2.428   Durbin-Watson:                   1.639
Prob(Omnibus):                  0.297   Jarque-Bera (JB):                2.082
Skew:                          -0.352   Prob(JB):                        0.353
Kurtosis:                       3.054   Cond. No.                         4.65
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.64 -> 2 => Автокорреляция отсутствует 

## log_income - Gender

#### Результаты анализа ANOVA

In [118]:
model3 =  ols("log_income ~ C(Gender)", df3).fit()
model3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.010
Method:                 Least Squares   F-statistic:                  0.004630
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.946
Time:                        00:54:34   Log-Likelihood:                -106.40
No. Observations:                 100   AIC:                             216.8
Df Residuals:                      98   BIC:                             222.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept            10.5533      0.163     64.940      0.000      10.231      10.876
C(Gender)[T.male]     0.0123      0.181      0.068      0.946      -0.346       0.371
==============================================================================
Omnibus:                        2.768   Durbin-Watson:                   1.812
Prob(Omnibus):                  0.251   Jarque-Bera (JB):                2.401
Skew:                          -0.378   Prob(JB):                        0.301
Kurtosis:                       3.064   Cond. No.                         4.39
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.81 -> 2 => Автокорреляция отсутствует 

## log_income - JobPref

#### Результаты анализа ANOVA

In [119]:
model4 =  ols("log_income ~ C(JobPref)", df3).fit()
model4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                 -0.013
Method:                 Least Squares   F-statistic:                    0.3674
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.693
Time:                        00:54:41   Log-Likelihood:                -106.03
No. Observations:                 100   AIC:                             218.1
Df Residuals:                      97   BIC:                             225.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
======================================================================================================================
                                                         coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------------
Intercept                                             10.5230      0.090    116.810      0.000      10.344      10.702
C(JobPref)[T.work for a multinational corporation]     0.1833      0.224      0.819      0.415      -0.261       0.627
C(JobPref)[T.work for a startup]                       0.0704      0.166      0.425      0.672      -0.258       0.399
==============================================================================
Omnibus:                        3.015   Durbin-Watson:                   1.825
Prob(Omnibus):                  0.221   Jarque-Bera (JB):                2.555
Skew:                          -0.386   Prob(JB):                        0.279
Kurtosis:                       3.133   Cond. No.                         3.46
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.83 -> 2 => Автокорреляция отсутствует 

## log_income - JobWherePref

#### Результаты анализа ANOVA

In [120]:
model5 =  ols("log_income ~ C(JobWherePref)", df3).fit()
model5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             log_income   R-squared:                       0.008
Model:                            OLS   Adj. R-squared:                 -0.013
Method:                 Least Squares   F-statistic:                    0.3802
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.685
Time:                        00:54:47   Log-Likelihood:                -106.02
No. Observations:                 100   AIC:                             218.0
Df Residuals:                      97   BIC:                             225.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=========================================================================================================================
                                                            coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------
Intercept                                                10.6470      0.122     87.532      0.000      10.406      10.888
C(JobWherePref)[T.in an office with other developers]    -0.1182      0.155     -0.761      0.448      -0.426       0.190
C(JobWherePref)[T.no preference]                         -0.1654      0.238     -0.695      0.489      -0.638       0.307
==============================================================================
Omnibus:                        2.239   Durbin-Watson:                   1.837
Prob(Omnibus):                  0.326   Jarque-Bera (JB):                1.902
Skew:                          -0.337   Prob(JB):                        0.386
Kurtosis:                       3.053   Cond. No.                         4.34
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.84 -> 2 => Автокорреляция отсутствует 

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.87 -> 2 => Автокорреляция отсутствует 

## Попробуем другую трансформацию

In [80]:
df2['sqrt_income'] = np.sqrt(df2.Income)
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.sqrt_income)
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.9709379076957703, p_value 1.1030138391196487e-12


#### Гипотеза H0 о нормальном распределении отклоняется p < 0.05

In [81]:
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.sqrt_income[:100])
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.9610716700553894, p_value 0.004798233043402433


#### Гипотеза H0 о нормальном распределении отклоняется p > 0.05

In [82]:
df2['inv_income'] = 1 / (df2.Income)
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.inv_income)
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.7213222980499268, p_value 1.0208670516310835e-36


#### Гипотеза H0 о нормальном распределении отклоняется p < 0.05

In [83]:
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.inv_income[:100])
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.7431211471557617, p_value 6.183916226309982e-12


#### Гипотеза H0 о нормальном распределении отклоняется p > 0.05

In [84]:
df2['sqrt3_income'] = df2.Income ** (1 / 3)
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.sqrt3_income)
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.9855682849884033, p_value 6.037085853449753e-08


#### Гипотеза H0 о нормальном распределении отклоняется p < 0.05

In [85]:
print(f"Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:")
z, p = sts.shapiro(df2.sqrt3_income[:100])
print(f"Статистика критерия {z}, p_value {p}")

Проверка гипотезы на нормальность распределения тестом Шапиро Уилка для уровня дохода:
Статистика критерия 0.979854166507721, p_value 0.12950226664543152


#### Гипотеза H0 о нормальном распределении принимается p > 0.05

## sqrt3_income - SchoolDegree

In [93]:
df3 = df2[:100]

#### Результаты анализа ANOVA

In [121]:
model6 =  ols("sqrt3_income ~ C(SchoolDegree)", df3).fit()
model6.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           sqrt3_income   R-squared:                       0.054
Model:                            OLS   Adj. R-squared:                 -0.041
Method:                 Least Squares   F-statistic:                    0.5709
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.818
Time:                        00:54:55   Log-Likelihood:                -345.46
No. Observations:                 100   AIC:                             710.9
Df Residuals:                      90   BIC:                             737.0
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                      35.6764      4.036      8.840      0.000      27.658      43.695
C(SchoolDegree)[T.associate's degree]                          -6.7914      5.059     -1.342      0.183     -16.843       3.260
C(SchoolDegree)[T.bachelor's degree]                            0.0291      4.260      0.007      0.995      -8.435       8.493
C(SchoolDegree)[T.high school diploma or equivalent (GED)]     -2.0106      4.943     -0.407      0.685     -11.831       7.810
C(SchoolDegree)[T.master's degree (non-professional)]          -0.8472      4.462     -0.190      0.850      -9.712       8.017
C(SchoolDegree)[T.no high school (secondary school)]            4.5308      9.025      0.502      0.617     -13.398      22.460
C(SchoolDegree)[T.professional degree (MBA, MD, JD, etc.)]     -0.1453      4.943     -0.029      0.977      -9.966       9.675
C(SchoolDegree)[T.some college credit, no degree]              -0.6204      4.660     -0.133      0.894      -9.879       8.638
C(SchoolDegree)[T.some high school]                            -3.0336      6.165     -0.492      0.624     -15.282       9.214
C(SchoolDegree)[T.trade, technical, or vocational training]    -0.8931      5.708     -0.156      0.876     -12.232      10.446
==============================================================================
Omnibus:                        2.985   Durbin-Watson:                   1.830
Prob(Omnibus):                  0.225   Jarque-Bera (JB):                2.361
Skew:                           0.344   Prob(JB):                        0.307
Kurtosis:                       3.307   Cond. No.                         18.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.87 -> 2 => Автокорреляция отсутствует 

## sqrt3_income - CityPopulation

#### Результаты анализа ANOVA

In [122]:
model7 =  ols("sqrt3_income ~ C(CityPopulation)", df3).fit()
model7.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           sqrt3_income   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                 -0.006
Method:                 Least Squares   F-statistic:                    0.7045
Date:                Fri, 02 Dec 2022   Prob (F-statistic):              0.497
Time:                        00:55:01   Log-Likelihood:                -347.52
No. Observations:                 100   AIC:                             701.0
Df Residuals:                      97   BIC:                             708.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                   35.5682      1.425     24.952      0.000      32.739      38.397
C(CityPopulation)[T.less than 100,000]      -0.1124      2.069     -0.054      0.957      -4.219       3.994
C(CityPopulation)[T.more than 1 million]    -1.9668      1.889     -1.041      0.300      -5.716       1.782
==============================================================================
Omnibus:                        4.732   Durbin-Watson:                   1.804
Prob(Omnibus):                  0.094   Jarque-Bera (JB):                4.072
Skew:                           0.416   Prob(JB):                        0.131
Kurtosis:                       3.534   Cond. No.                         3.89
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### p(Omnibus) > 0.05 => Наличие гомоскедастичности 
#### DW = 1.80 -> 2 => Автокорреляция отсутствует 

# Задание 5

## Интерпретация результатов

In [123]:
anova = sm.stats.anova_lm(model, typ=1)
anova

df     sum_sq   mean_sq         F   PR(>F)
C(SchoolDegree)   9.0   3.066072  0.340675  0.664936  0.73834
Residual         90.0  46.110789  0.512342       NaN      NaN

#### Вывод: Доход от уровня образования не зависит

In [127]:
anova = sm.stats.anova_lm(model1, typ=1)
anova

df     sum_sq   mean_sq         F    PR(>F)
C(CityPopulation)   2.0   1.062998  0.531499  1.071529  0.346503
Residual           97.0  48.113863  0.496019       NaN       NaN

#### Вывод: Доход от уровня населения города не зависит

In [128]:
anova = sm.stats.anova_lm(model2, typ=1)
anova

df     sum_sq   mean_sq         F   PR(>F)
C(EmploymentStatus)   2.0   2.141799  1.070899  2.208506  0.11536
Residual             97.0  47.035063  0.484898       NaN      NaN

#### Вывод: Доход от статуса работника не зависит

In [129]:
anova = sm.stats.anova_lm(model3, typ=1)
anova

df     sum_sq   mean_sq        F   PR(>F)
C(Gender)   1.0   0.002323  0.002323  0.00463  0.94589
Residual   98.0  49.174538  0.501781      NaN      NaN

#### Вывод: Доход от пола не зависит

In [130]:
anova = sm.stats.anova_lm(model4, typ=1)
anova

df     sum_sq   mean_sq         F    PR(>F)
C(JobPref)   2.0   0.369769  0.184884  0.367442  0.693463
Residual    97.0  48.807092  0.503166       NaN       NaN

#### Вывод: Доход от предпочтения к работе не зависит

In [132]:
anova = sm.stats.anova_lm(model5, typ=1)
anova

df     sum_sq   mean_sq        F    PR(>F)
C(JobWherePref)   2.0   0.382537  0.191269  0.38023  0.684719
Residual         97.0  48.794324  0.503034      NaN       NaN

#### Вывод: Доход от места работы не зависит